File: module/vm/wxvm/xvm/runtime/wasi/resolver.go
/*
 * Copyright (C) BABEC. All rights reserved.
 * Copyright (C) THL A29 Limited, a Tencent company. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

package wasi

import "chainmaker.org/chainmaker-go/wxvm/xvm/exec"

var resolver = exec.MapResolver(map[string]interface{}{
	"env.___wasi_fd_prestat_get": func(ctx exec.Context, x, y uint32) uint32 {
		return 8
	},
	"env.___wasi_fd_fdstat_get": func(ctx exec.Context, x, y uint32) uint32 {
		return 8
	},
	"env.___wasi_fd_prestat_dir_name": func(ctx exec.Context, x, y, z uint32) uint32 {
		return 8
	},
	"env.___wasi_fd_close": func(ctx exec.Context, x uint32) uint32 {
		return 8
	},
	"env.___wasi_fd_seek": func(ctx exec.Context, x, y, z, w uint32) uint32 {
		return 8
	},
	"env.___wasi_fd_write": func(ctx exec.Context, x, y, z, w uint32) uint32 {
		return 8
	},
	"env.___wasi_environ_sizes_get": func(ctx exec.Context, x, y uint32) uint32 {
		return 0
	},
	"env.___wasi_environ_get": func(ctx exec.Context, x, y uint32) uint32 {
		return 0
	},
	"env.___wasi_args_sizes_get": func(ctx exec.Context, x, y uint32) uint32 {
		return 0
	},
	"env.___wasi_args_get": func(ctx exec.Context, x, y uint32) uint32 {
		return 0
	},
	"env.___wasi_proc_exit": func(ctx exec.Context, x uint32) uint32 {
		exec.Throw(exec.NewTrap("exit"))
		return 0
	},
})

// NewResolver return exec.Resolver which resolves symbols needed by wasi environment
func NewResolver() exec.Resolver {
	return resolver
}

File: module/vm/wxvm/xvm/exec/resolver.c
/*
 * Copyright (C) BABEC. All rights reserved.
 * Copyright (C) THL A29 Limited, a Tencent company. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

#include "wxvm.h"

extern wasm_rt_func_handle_t wxvm_resolve_func(void* env, char* module, char* name);
extern int64_t wxvm_resolve_global(void* env, char* module, char* name);
extern uint32_t wxvm_call_func(void* env, wasm_rt_func_handle_t handle, wxvm_context_t* ctx, uint32_t* params, uint32_t param_len);

wxvm_resolver_t make_resolver_t(void* env) {
	wxvm_resolver_t r;
	r.env = env;
	r.resolve_func = wxvm_resolve_func;
	r.resolve_global = wxvm_resolve_global;
	r.call_func = wxvm_call_func;
	return r;
}
